File: lua/vectors.lua
printf = function ( ... ) print(string.format(...)) end


function vectable( runs )
	local test = {}

	for i = 1, 1024 do
		local x, y, z, w = math.random(), math.random(), math.random(), math.random()
		test[#test+1] = { x = x, y = y, z = z, w = w }
	end

	local start = os.time()

	local total = 0

	for i = 1, runs do
		for j = 1, #test-1 do
			local a, b = test[j], test[j+1]
			local dot = (a.x * b.x) + (a.y * b.y) + (a.z * b.z) + (a.w * b.w)
			total = total + dot
		end
	end

	local finish = os.time()

	printf('table %.2fs total:%f', os.difftime(finish, start), total)
end

function vecarray( runs )
	local test = {}

	for i = 1, 1024 do
		local x, y, z, w = math.random(), math.random(), math.random(), math.random()
		test[#test+1] = { x, y, z, w }
	end

	local start = os.time()

	local total = 0

	for i = 1, runs do
		for j = 1, #test-1 do
			local a, b = test[j], test[j+1]
			local dot = (a[1] * b[1]) + (a[2] * b[2]) + (a[3] * b[3]) + (a[4] * b[4])
			total = total + dot
		end
	end

	local finish = os.time()

	printf('array %.2fs total:%f', os.difftime(finish, start), total)
end


local runs = 3000
printf('#runs:%d', runs)
for i = 1, 5 do
	vectable(runs)
	vecarray(runs)
end

File: c++/vectors.cpp
#include <stdio.h>

#include <stdlib.h>
#include <time.h>




struct Vector {
	float x;
	float y;
	float z;
	float w;
};

float rand_float()
{
	// 0.0 to 1.0
	return float((float)rand() / (float)RAND_MAX );
}

void vectable(unsigned int runs)
{
	
	Vector data[1024];

	for (int i = 0; i < 1024; ++i) {
		data[i].x = rand_float();
		data[i].y = rand_float();
		data[i].z = rand_float();
		data[i].w = rand_float();
	}

	int start = clock();

	
	double total = 0;
	double dot = 0;

	for (unsigned int i = 0; i < runs; ++i) {
		for (int j = 0; j < 1023; ++j) {
			dot = (data[j].x * data[j+1].x) + (data[j].y * data[j+1].y) + (data[j].z * data[j+1].z) + (data[j].w * data[j+1].w);
			total += dot;
		}
	}
	int end = clock();

	int cycles = end-start;

	float seconds = (double)cycles / CLOCKS_PER_SEC;
	printf("took %f seconds.  total: %f\n", seconds, total);
}



int main(int argc, char *argv[])
{
    
	unsigned int runs = 20000;
	printf("Runs: %i \n", runs);
	for (int i = 0; i < 5; ++i) {
		vectable(runs);
	}
	return 0;
}